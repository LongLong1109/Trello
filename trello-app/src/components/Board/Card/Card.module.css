.card {
  border-width: 2px;
  border-color: transparent;

  &:hover {
    cursor: pointer;
    border-color: var(--mantine-color-backgrounds-2);
    border-width: 2px;
  }
}

.checkbox {
  border-width: 0;
  font-weight: 500;
  cursor: pointer;
}
