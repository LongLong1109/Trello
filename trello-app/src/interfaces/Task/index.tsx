export interface Task {
  id: number
  title: string
  labels?: string[]
  description?: string
  dueDate?: Date | null
  comment?: string
  orderId?: number
}

export interface ListTask {
  id: number
  name: string
  tasks: Task[]
}

export interface TaskListItem extends Task {
  listId: number
}

export interface TasksResponse {
  data: Task[] | undefined
}
