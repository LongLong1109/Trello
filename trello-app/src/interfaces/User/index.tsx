export interface UserLogin {
  email: string
  password: string
}

export interface UserRegister extends UserLogin {
  firstName: string
  lastName: string
}

export interface BaseUserInfo extends UserRegister {
  id: string
}

export interface UserResponse {
  accessToken: string
  user: BaseUserInfo
}

export interface UserAction {
  userAuth: UserResponse | null
  register: (data: UserRegister) => Promise<UserResponse>
  login: (data: UserLogin) => Promise<UserResponse>
  logout: () => Promise<void>
}
