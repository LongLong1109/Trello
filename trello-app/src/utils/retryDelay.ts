export const retryDelay = (attemptIndex: number) => Math.min(1000 * 2 ** attemptIndex, 30000)
