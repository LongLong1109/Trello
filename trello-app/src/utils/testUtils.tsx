import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { MantineProvider } from '@mantine/core'
const AllBrowserRouter = ({ children }: { children: React.ReactNode }) => {
  return <MantineProvider>{children}</MantineProvider>
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllBrowserRouter, ...options })

export * from '@testing-library/react'
export * from '@testing-library/user-event'
export { customRender as render }
