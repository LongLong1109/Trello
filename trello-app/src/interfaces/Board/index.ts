import { ListTask } from '../Task'

export interface UserBoard {
  columnId?: string
  columns: ListTask[]
}

export interface ColumnItem {
  id: number
  name: string
  order: number
}

export interface Columns {
  columnId: string
  columns: ColumnItem[]
}

export interface ColumnsResponse {
  data: Columns[]
}
