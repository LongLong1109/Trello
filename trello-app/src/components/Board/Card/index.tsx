import React from 'react'
import { Card, Text, Paper, Flex, Badge } from '@mantine/core'
import { IconMessagePlus, IconFileDescription } from '@tabler/icons-react'
// interface
import { Task } from '@/interfaces/Task'

// utils
import { isDateOverdue } from '@/utils/validateDate'

interface TaskCardProps {
  task: Task
  onDragStart: (event: React.DragEvent<HTMLDivElement>, taskId: number) => void
  onOpenCard: (value: Task) => void
}

const TaskCard = ({ task, onDragStart, onOpenCard }: TaskCardProps) => {
  const { id, title, labels, description, comment, dueDate } = task
  const taskDueDate = new Date(dueDate ?? '')
  const checkDateOverdue = isDateOverdue(taskDueDate)

  const handleCardClick = () => {
    onOpenCard(task)
  }

  return (
    <Card
      data-testid='card-id'
      shadow='md'
      padding='lg'
      mb='10'
      draggable
      onDragStart={(event) => onDragStart(event, id)}
      onClick={handleCardClick}
    >
      <Text fs='500' pb='10'>
        {title}
      </Text>
      {labels && (
        <Flex gap='5'>
          {labels.map((label) => (
            <Paper w='30' h='10' radius='4' bg={label} key={label} />
          ))}
        </Flex>
      )}
      <Flex gap='10' align='center' mt='10'>
        {dueDate && (
          <Badge variant='light' color={checkDateOverdue ? 'red' : 'blue'} radius='xs'>
            {taskDueDate.toLocaleDateString('en-US', { day: '2-digit', month: '2-digit' })}
          </Badge>
        )}
        {description && <IconFileDescription size='14' />}
        {comment && <IconMessagePlus size='14' />}
      </Flex>
    </Card>
  )
}

export default TaskCard
