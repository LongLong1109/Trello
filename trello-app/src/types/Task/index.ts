import { Task } from '@/interfaces/Task'
import { DatesRangeValue } from '@mantine/dates'
import { CheckboxItem } from '@/components/common'

export type TaskPayload = Omit<Task, 'id'>

export enum TASK_PROPERTY_KEY {
  DUE_DATE = 'dateRange',
  DESCRIPTION = 'description',
  COMMENT = 'comment',
  LABELS = 'labels',
  CHECKED = 'checked',
}

export type TaskPropertyValue = DatesRangeValue | string | CheckboxItem[] | boolean
