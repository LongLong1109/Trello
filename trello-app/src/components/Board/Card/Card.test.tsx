// component
import Card from '.';

// interface
import { Task } from '@/interfaces/Task';

// utils
import { render, screen, fireEvent } from '@/utils/testUtils';

describe('Card component', () => {
  const mockTask: Task = {
    id: 1,
    title: 'Test Task',
    labels: ['Label1', 'Label2'],
    description: 'This is a test task description',
    dueDate: new Date('2023-12-31'),
  };

  const mockOnDragStart = jest.fn();
  const mockOnOpenCard = jest.fn();

  const renderTaskCard = () => {
    return render(
      <Card task={mockTask} onDragStart={mockOnDragStart} onOpenCard={mockOnOpenCard} />,
    );
  };

  it('should be matches snapshot', () => {
    const { asFragment } = renderTaskCard();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the task card with correct content', () => {
    renderTaskCard();

    expect(screen.getByText('Test Task')).toBeInTheDocument();
    expect(screen.getByText('Label1, Label2')).toBeInTheDocument();
    expect(screen.getByText('This is a test task description')).toBeInTheDocument();
    expect(screen.getByText('12/31/2023')).toBeInTheDocument();
  });

  it('should call onDragStart with the correct arguments', () => {
    renderTaskCard();
    const card = screen.getByTestId('card-id');
    fireEvent.dragStart(card);

    expect(mockOnDragStart).toHaveBeenCalledTimes(1);
    expect(mockOnDragStart).toHaveBeenCalledWith(expect.any(Object), 1);
  });

  it('should call onClick with the correct arguments', () => {
    renderTaskCard();
    const card = screen.getByTestId('card-id');
    fireEvent.click(card);

    expect(mockOnOpenCard).toHaveBeenCalledTimes(1);
  });
});
