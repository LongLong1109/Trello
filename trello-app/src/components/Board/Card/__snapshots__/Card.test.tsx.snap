// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should be matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_e615b15f mantine-Card-root m_1b7284a3 mantine-Paper-root"
    draggable="true"
    style="--paper-shadow: var(--mantine-shadow-md); --card-padding: var(--mantine-spacing-lg);"
  >
    <p
      class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
      style="font-style: 500;"
    >
      Test Task
    </p>
    <p
      class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    >
      Label1, Label2
    </p>
    <p
      class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    >
      This is a test task description
    </p>
    <p
      class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    >
      12/31/2023
    </p>
  </div>
</DocumentFragment>
`;
